### PR TITLE
Fix positioning monthly tickformat when initial auto dtick is weekly

### LINKED ISSUE #5207: period label positions off with weekly dtick and monthly tickformat
[Bug demo](https://codepen.io/MojtabaSamimi/pen/oNLgVZY?editors=0010)

### PR DESCRIPTION
**Pull request by archmoj**
Fixes #5207.
Demo: [Before](https://codepen.io/MojtabaSamimi/pen/oNLgVZY?editors=0010) vs [After](https://codepen.io/MojtabaSamimi/pen/qBNEvYR?editors=1000)

@plotly/plotly_js

### CODE REVIEW

File: src/plots/cartesian/axes.js [hunk 1]
```diff
@@ -706,11 +707,16 @@ axes.calcTicks = function calcTicks(ax, opts) {
                 ) ax.dtick = 'M12';
             }
         }
+
+        if(prevDtick !== ax.dtick) {
+            // move tick0 back
+            ax.tick0 = axes.tickIncrement(ax.tick0, prevDtick, !axrev, ax.calendar);
```
Comment (alexcjohnson) [inline src/plots/cartesian/axes.js:713]: This seems problematic - if you set a specific `tick0` it won't be honored? In any event I don't think we can be pushing a different value back into `ax` at this point, but if `tickFirst` needs a different effective start point perhaps we could make it an option to `tickFirst`?

Comment (archmoj) [reply src/plots/cartesian/axes.js:715]: Good call. Addressed in a4cb02b.

**Comment by alexcjohnson** [comment]
I don't really understand what's going on here, but strange things happen when I start interacting with this graph - a tiny drag and the ticks jump back to the incorrect "before" position. I think we need to resolve this at an earlier stage in the process.

**Comment by archmoj** [comment]
> I don't really understand what's going on here, but strange things happen when I start interacting with this graph - a tiny drag and the ticks jump back to the incorrect "before" position. I think we need to resolve this at an earlier stage in the process.

Good catch. In fact it was previous commit a4cb02b that contributed in this bug.
For now I reverted part of it in ecb5067. 
Now investigating if we could possibly resolve it earlier in the process.
Thanks.

**Comment by archmoj** [comment]
@alexcjohnson please note that the `tick0` tweaking part was removed in https://github.com/plotly/plotly.js/pull/5208/commits/5d61cef1140b7de94858e339cabb808e7e7df6f6.

Then commit https://github.com/plotly/plotly.js/pull/5208/commits/77a51d34b3f5e0b06fed94e61758df4225f210da was enough to fixe the bug.

And https://github.com/plotly/plotly.js/pull/5208/commits/c42358c128c7ea2b44b2531052ea9ac3f877aaf4 as well as https://github.com/plotly/plotly.js/pull/5208/commits/b78a80bb03950b30f18e0e01a3cae442ddc0df6c basically moved most of the `period` logic into functions i.e. to help minimize the `calcTicks` function.

**Comment by alexcjohnson** [comment]
Much cleaner! But there's still one issue here, because the auto `dtick` prior to considering `tickformat` is weeks, the auto `tick0` is `'2000-01-02'`, hence the tick you see is on Jan 2. But then when you zoom out so the auto `dtick` is months, `tick0` and the ticks you see jump back to the first of the month.

Seems to me like either `adjustPeriodDelta` needs to also adjust `tick0`, or perhaps even cleaner, the whole `adjustPeriodDelta` logic could be pushed into `axes.autoTicks`?

**Comment by archmoj** [comment]
> Much cleaner! But there's still one issue here, because the auto `dtick` prior to considering `tickformat` is weeks, the auto `tick0` is `'2000-01-02'`, hence the tick you see is on Jan 2. But then when you zoom out so the auto `dtick` is months, `tick0` and the ticks you see jump back to the first of the month.
> 
> Seems to me like either `adjustPeriodDelta` needs to also adjust `tick0`, or perhaps even cleaner, the whole `adjustPeriodDelta` logic could be pushed into `axes.autoTicks`?

Good eye.

**Review by alexcjohnson — APPROVED** [review]
💃  🎯

**Pull request merged**